pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -76,30 +73,43 @@ func WithCustomMapper(newMapper func() (meta.RESTMapper, error)) DynamicRESTMapp
 // RESTMapper dynamically discovers resource types at runtime. opts
 // configure the RESTMapper.
 func NewDynamicRESTMapper(cfg *rest.Config, opts ...DynamicRESTMapperOption) (meta.RESTMapper, error) {
-	client, err := discovery.NewDiscoveryClientForConfig(cfg)
-	if err != nil {
-		return nil, err
-	}
 	drm := &dynamicRESTMapper{
 		limiter: rate.NewLimiter(rate.Limit(defaultRefillRate), defaultLimitSize),
-		newMapper: func() (meta.RESTMapper, error) {
-			groupResources, err := restmapper.GetAPIGroupResources(client)
-			if err != nil {
-				return nil, err
-			}
-			return restmapper.NewDiscoveryRESTMapper(groupResources), nil
-		},
 	}
 	for _, opt := range opts {
-		if err = opt(drm); err != nil {
+		if err := opt(drm); err != nil {
 			return nil, err
 		}
 	}
-	if !drm.lazy {
-		if err := drm.setStaticMapper(); err != nil {
+
+	if drm.newMapper == nil {
+		// If custom mapper is not set, we wse NewDeferredDiscoveryRESTMapper for lazy loading
+		// or standard NewDiscoveryRESTMapper for other cases.
+		client, err := discovery.NewDiscoveryClientForConfig(cfg)
+		if err != nil {
 			return nil, err
 		}
+
+		if !drm.lazy {
+			drm.newMapper = func() (meta.RESTMapper, error) {
+				groupResources, err := restmapper.GetAPIGroupResources(client)
+				if err != nil {
+					return nil, err
+				}
+
+				return restmapper.NewDiscoveryRESTMapper(groupResources), nil
+			}
+		} else {
+			drm.newMapper = func() (meta.RESTMapper, error) {
+				return restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(client)), nil
```

Comment:
From what I can tell this only defers until the first usage and then fetches all groups, rather than the ones that are needed:
1. `RESTMapping` constructs a `delegate`
2. `getDelegate` calls `GetAPIGroupResources`
3. `GetAPIGroupResources` fetches all apis

Let me know if I missed something.